test/integration/garbagecollector/garbage_collector_test.go

```diff
@@ -1017,3 +1021,94 @@ func TestCRDDeletionCascading(t *testing.T) {
 		t.Fatalf("failed waiting for dependent %q (owned by %q) to be deleted", dependent.GetName(), owner.GetName())
 	}
 }
+
+func TestReplicaSetsAppsV1DefaultGCPolicy(t *testing.T) {
+	ctx := setup(t, 5)
+	defer ctx.tearDown()
+
+	clientSet := ctx.clientSet
+
+	ns := createNamespaceOrDie("default-gc-policy", clientSet, t)
+	defer deleteNamespaceOrDie(ns.Name, clientSet, t)
+
+	// apps/v1 replicaset
+	rsName := "gc-rs-appsv1"
+	finalizers := []string{"kube.io/dummy-finalizer"}
+	rs := &appsv1.ReplicaSet{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ReplicaSet",
+			APIVersion: "apps/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    ns.Name,
+			GenerateName: "gc-rs-test-appsv1",
+			Name:         rsName,
+			Finalizers:   finalizers,
+		},
+		Spec: appsv1.ReplicaSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{"name": "test-appsv1"},
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"name": "test-appsv1"},
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name:  "fake-name",
+							Image: "fakeimage",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	rsV1Client := clientSet.AppsV1().ReplicaSets(ns.Name)
+	rs, err := rsV1Client.Create(rs)
+	if err != nil {
+		t.Fatalf("Failed to create rs: %v", err)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		w, err := clientSet.AppsV1().RESTClient().Get().
+			Namespace(ns.Name).
+			Resource("replicasets").
+			Name(rsName).
+			Watch()
```

Comment:
What we expect is for no finalizers to get added. I think we can make this test simpler and more reliable by doing something like this, instead of using a watch:

1. When creating the RS, include our own fake finalizer like `example.com/fake`.
2. Delete the RS.
3. Poll the RS until it has a non-nil `DeletionTimestamp`, and then verify that the only finalizer present is `example.com/fake`.
4. Update the RS to remove the fake finalizer, then delete it for real.

This pattern should work for the other test too, except you verify that the orphan finalizer gets added.

Reply:
Thanks @enisoc